### PR TITLE
Upgrade to python3

### LINKED ISSUE #58: Migrate to Django 2.1
Similar issue #24 

Upgrade the codebase to Python 3 and Django 2.1.
Some work is already done in https://github.com/spdx/spdx-online-tools/tree/python3

### PR DESCRIPTION
**Pull request by rtgdk**
Updated python3 branch with master and made changes to support python3.7+. Able to run the server, some tests are failing locally either due to bytes/str conversion or jpype upgrade.

Main changes done in this PR:

1. Updated python3 branch with master.
2. Updated requirements.txt - all packages are updated to a newer version with python3.7+ support.
3. jpype changed JavaException to JException.
4. Python3 related changes - 
      * izip changed to zip
      *  relative imports
      * Use wb instead of w when opening files - to write bytes content
5. Django2 related changes -
     * is_authenticated() changed to is_authenticated
     * on_delete=models.CASCADE added to foreign keys
     * Fixed example dir path by settings up EXAMPLES_DIR variable in setting.py
6. Ran 2to3 on the whole code
7. Updated tests to use bytes when comparison
8. Update dockerfiles, github workflow for py3
9. Updated Readme.md for py3 install and run instructions (Major change is to remove virtualenv install since python3 supports virtual env itself with `python -m venv`

Major Issue Left to Address:
1. Python3 changed how we used to represent in strings in python2. Now HttpResponse and opening/writing file uses bytes instead of string- https://docs.python.org/3.0/whatsnew/3.0.html#text-vs-data-instead-of-unicode-vs-8-bit
This issue made a lot of tests fail due to comparision b/w str and bytes. Most of them are fixed by changing expected comparison to bytes by adding b in the str. Although, more robust changes might be required to come up with a way to compare actual output - either changing the response to str by encoding or str to byted by decoding. `(encode()/decode() functions of str)`

 ----- 
Some of the changes are picked from https://github.com/spdx/spdx-online-tools/pull/178, although it was not updated with master and did not have the full changes, so had to create a new PR. Thanks @victorphoenix3 for all the work.
 
----
**To review this PR, either checkout the branch locally and run the server or check the git diff with "master" instead. I'll open a new PR from python3->master once this gets merged with all failures resolved.**

Fixes #58 #24

### CODE REVIEW

**Comment by rtgdk** [comment]
Would appreciate your review and local testing
cc @goneall  @techytushar  @Ugtan  @kmehant

**Comment by goneall** [comment]
@rtgdk Could you add/update the dockerfile configurations?  I'll do some local tests using Docker now that I have the production infrastructure (mostly) working.

**Comment by rtgdk** [comment]
@goneall The docker files are updated with master, do you want me to update them with your PR #212 changes ?

**Comment by rtgdk** [comment]
Also, the server runs fine but some tests are failing due to reasons mentioned in the description (mostly because of the difference b/w bytes and str in python3). So, you can expect some functionalities to not work fully.

**Comment by goneall** [comment]
I'm getting errors bringing up the server in the spdx-license-matcher:

```

EXITED (1)





Traceback (most recent call last):

File "src/manage.py", line 22, in <module>

execute_from_command_line(sys.argv)

File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line

utility.execute()

File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 375, in execute

self.fetch_command(subcommand).run_from_argv(self.argv)

File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 323, in run_from_argv

self.execute(*args, **cmd_options)

File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 361, in execute

self.check()

File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 387, in check

all_issues = self._run_checks(

File "/usr/local/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 65, in _run_checks

issues.extend(super()._run_checks(**kwargs))

File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 377, in _run_checks

return checks.run_checks(**kwargs)

File "/usr/local/lib/python3.8/site-packages/django/core/checks/registry.py", line 72, in run_checks

new_errors = check(app_configs=app_configs)

File "/usr/local/lib/python3.8/site-packages/django/core/checks/urls.py", line 40, in check_url_namespaces_unique

all_namespaces = _load_all_namespaces(resolver)

File "/usr/local/lib/python3.8/site-packages/django/core/checks/urls.py", line 57, in _load_all_namespaces

url_patterns = getattr(resolver, 'url_patterns', [])

File "/usr/local/lib/python3.8/site-packages/django/utils/functional.py", line 80, in __get__

res = instance.__dict__[self.name] = self.func(instance)

File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 584, in url_patterns

patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)

File "/usr/local/lib/python3.8/site-packages/django/utils/functional.py", line 80, in __get__

res = instance.__dict__[self.name] = self.func(instance)

File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 577, in urlconf_module

return import_module(self.urlconf_name)

File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module

return _bootstrap._gcd_import(name[level:], package, level)

File "<frozen importlib._bootstrap>", line 1014, in _gcd_import

File "<frozen importlib._bootstrap>", line 991, in _find_and_load

File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked

File "<frozen importlib._bootstrap>", line 671, in _load_unlocked

File "<frozen importlib._bootstrap_external>", line 783, in exec_module

File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed

File "/spdxonlinetools/src/src/urls.py", line 49, in <module>

path('app/', include('app.urls')),

File "/usr/local/lib/python3.8/site-packages/django/urls/conf.py", line 34, in include

urlconf_module = import_module(urlconf_module)

File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module

return _bootstrap._gcd_import(name[level:], package, level)

File "<frozen importlib._bootstrap>", line 1014, in _gcd_import

File "<frozen importlib._bootstrap>", line 991, in _find_and_load

File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked

File "<frozen importlib._bootstrap>", line 671, in _load_unlocked

File "<frozen importlib._bootstrap_external>", line 783, in exec_module

File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed

File "/spdxonlinetools/src/app/urls.py", line 19, in <module>

from app import views

File "/spdxonlinetools/src/app/views.py", line 51, in <module>

import app.utils as utils

File "/spdxonlinetools/src/app/utils.py", line 24, in <module>

from spdx_license_matcher.build_licenses import build_spdx_licenses

File "/src/spdx-license-matcher/spdx_license_matcher/build_licenses.py", line 8, in <module>

from normalize import normalize

ModuleNotFoundError: No module named 'normalize'


```

**Comment by rtgdk** [comment]
Looks like a python3 relative import issue with https://github.com/spdx/spdx-license-matcher. I think license matcher was written in python2.  @Ugtan  can you confirm? python3 changed how we used to import relative modules/files.

**Comment by Ugtan** [comment]
@rtgdk Yeah. The license matcher is written in python2. Do you want it to be compatible with both python2 and 3 or if we are planning to drop support for spdx online tools python2, we could simply drop it from the license matcher too. Let me know your thoughts. I will update the license matcher accordingly.

**Comment by rtgdk** [comment]
@Ugtan I think we can have a separate branch `python3` in  https://github.com/spdx/spdx-license-matcher which is compatible with python3 and later on merge this with master. Maybe when we merge it, we can move the python2 code to another branch - `python2` so that we preserve that state for future.


I'll update the requirements for this branch and point it to spdx-license-matcher python3 branch.

**Comment by goneall** [comment]
I reviewed the changes and only found a couple very minor documentation related items.  I'll add a PR to this branch with suggested changes.

**Comment by goneall** [comment]
Trying this out on Windows using Python version 3.8, I'm getting the following error trying to import the spdx_license_matcher functions:

```
  File "C:\Users\gary\git\temp\spdx-online-tools\src\app\utils.py", line 24, in <module>
    from spdx_license_matcher.build_licenses import build_spdx_licenses
  File "c:\users\gary\git\temp\spdx-online-tools\venv\src\spdx-license-matcher\spdx_license_matcher\build_licenses.py", line 8, in <module>
    from normalize import normalize
ModuleNotFoundError: No module named 'normalize'
```

Note that I'm getting the following errors during the requirements installatin which are referenced in https://github.com/spdx/spdx-license-matcher/issues/5 

```
ERROR: spdx-license-matcher 0.0.1 has requirement jpype1==0.6.3, but you'll have jpype1 1.0.1 which is incompatible.
ERROR: spdx-license-matcher 0.0.1 has requirement python-dotenv==0.13.0, but you'll have python-dotenv 0.14.0 which is incompatible.
ERROR: spdx-license-matcher 0.0.1 has requirement requests==2.22.0, but you'll have requests 2.24.0 which is incompatible.
```

**Comment by goneall** [comment]
@rtgdk Could you merge in the last few commits from master?  I don't think they will conflict - it's just a few fixes to the production scripts.

**Comment by goneall** [comment]
I ran into the same error `ModuleNotFoundError: No module named 'normalize'` using the Docker image.

**Comment by rtgdk** [comment]
@goneall Apologies for not mentioning earlier. Yeah, this error occurs since [spdx-license-matcher](https://github.com/spdx/spdx-license-matcher) is not updated to python3 (relative imports doesn't work as it used to). I had raised a PR - https://github.com/spdx/spdx-license-matcher/pull/12 for that. Turns out I don't have merge permission for that.  For my local development, I switched to that branch and installed it manually (python setup.py install). I don't have write access to merge the PR. Also, once merged we need to create a new release v2.0 and update requirements.txt. 
If you have the repo setup locally, then just pull the branch `rtgdk:python3` and run python setup.py install after activating the virtualenv.

**Comment by rtgdk** [comment]
@goneall I have merged master. [This](https://github.com/spdx/spdx-online-tools/pull/268/files#diff-c0ab48db8c0ba141d9ab4c3925127be58b47bc5aae822d8950b8bd10743c4b3bR507) part is not needed in python3 from the [encoding PR](https://github.com/spdx/spdx-online-tools/pull/268/).
Also, have created a new release v2.0 for license-matcher and updated requirements. Let me know, if you face the above import error again.

**Comment by goneall** [comment]
Thanks @rtgdk  - This resolved the issue.

I'm still doing some testing - I found another `str.decode(...` call which was causing an issue in license submissions.  I created a PR to resolve.

I'll do some more testing tomorrow - so far, looking pretty good 👍

**Comment by goneall** [comment]
Just a note on unit test failures - these may be the same as mentioned at the start of the PR - just adding a few details.

Here's some of the specifics on the unit test failures:

```
Traceback (most recent call last):
  File "C:\Users\gary\git\temp\spdx-online-tools\src\api\tests.py", line 182, in test_convert_xlsxtordf_api
    resp = self.client.post(reverse("convert-api"),{"file":self.xlsx_file,"from_format":self.xlsx,"to_format":self.rdf,"cfilename":"xlsxtordf-apitest"},format="multipart")
  File "C:\Users\gary\git\temp\spdx-online-tools\venv\lib\site-packages\rest_framework\test.py", line 293, in post
    response = super().post(
  File "C:\Users\gary\git\temp\spdx-online-tools\venv\lib\site-packages\rest_framework\test.py", line 206, in post
    data, content_type = self._encode_data(data, format, content_type)
  File "C:\Users\gary\git\temp\spdx-online-tools\venv\lib\site-packages\rest_framework\test.py", line 179, in _encode_data
    ret = renderer.render(data)
  File "C:\Users\gary\git\temp\spdx-online-tools\venv\lib\site-packages\rest_framework\renderers.py", line 914, in render
    return encode_multipart(self.BOUNDARY, data)
  File "C:\Users\gary\git\temp\spdx-online-tools\venv\lib\site-packages\django\test\client.py", line 201, in encode_multipart
    lines.extend(encode_file(boundary, key, value))
  File "C:\Users\gary\git\temp\spdx-online-tools\venv\lib\site-packages\django\test\client.py", line 253, in encode_file
    to_bytes(file.read())
  File "C:\Program Files\Python38\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x8d in position 1786: character maps to <undefined>
```

Similar errors occurs for several tests.

```
TypeError: a bytes-like object is required, not 'str'
```
error occurs in several test case files.

```
Traceback (most recent call last):
  File "C:\Users\gary\git\temp\spdx-online-tools\venv\lib\site-packages\selenium\webdriver\common\service.py", line 72, in start
    self.process = subprocess.Popen(cmd, env=self.env,
  File "C:\Program Files\Python38\lib\subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\Python38\lib\subprocess.py", line 1307, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
```

occurs in several test cases - this may be a missing configuration on my part

```
======================================================================
ERROR: test_generate_xml (api.tests.SubmitLicenseModelsTests)
View for generating an xml from license submittal form fields
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\gary\git\temp\spdx-online-tools\src\api\tests.py", line 390, in test_generate_xml
    xml = generateLicenseXml(self.osiApproved, self.shortIdentifier, self.fullname, self.listVersionAdded,
  File "C:\Users\gary\git\temp\spdx-online-tools\src\api\views.py", line 613, in generateLicenseXml
    xmlString = ET.tostring(root, method='xml').replace('>','>\n')
TypeError: a bytes-like object is required, not 'str'
```

**Comment by goneall** [comment]
Running into an issue submitting a new license: 

```
Unexpected error, please email the SPDX technical workgroup that the following error has occurred: Traceback (most recent call last): File "/spdxonlinetools/src/app/views.py", line 152, in submitNewLicense matches, issueUrl = utils.check_new_licenses_and_rejected_licenses(licenseText, urlType) File "/spdxonlinetools/src/app/utils.py", line 497, in check_new_licenses_and_rejected_licenses licenseData = get_license_data(issues) File "/spdxonlinetools/src/app/utils.py", line 473, in get_license_data licenseText = parseXmlString(licenseXml)['text'] File "/spdxonlinetools/src/app/utils.py", line 373, in parseXmlString tree = ET.ElementTree(ET.fromstring(xmlString)) File "/usr/local/lib/python3.8/xml/etree/ElementTree.py", line 1320, in XML parser.feed(text) File "<string>", line None xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 1, column 1
```

The string being passed into xml.etree.ElementTree.ParseError is:

```
'b\'<SPDXLicenseCollection xmlns="http://www.spdx.org/license"><license isOsiApproved="true" licenseId="testid" name="Test License [Please Ignore]" listVersionAdded=""><crossRefs><crossRef>http://source1.com</crossRef></crossRefs><standardLicenseHeader /><notes /><text><p>test not there yet</p></text></license></SPDXLicenseCollection>\''
```

This is fetched from the database from the following code in utils.py:

```
licenseXml = str(LicenseRequest.objects.get(id=dbId, shortIdentifier=licenseIdentifier).xml)
```

@rtgdk @Ugtan  let me know if you have any ideas - looks like something related to string encoding.

**Comment by rtgdk** [comment]
@goneall Regarding the test failures - yes mostly they are related to encoding difference b/w bytes and str. I had to make many forceful string to bytes conversion in tests in `src/app/tests.py` file to make them pass since our Django server API code is returning bytes output.

Regarding the `xml.etree.ElementTree.ParseError` error - again this is due to incorrect conversion of bytes to str. 

image

Line 472: utils.py - `licenseXml = str(LicenseRequest.objects.get(id=dbId, shortIdentifier=licenseIdentifier).xml)` causes this incorrect conversion.  So in our database, some row data (xml column) is saved in bytes instead of string. We can confirm this by looking at the dbId this license submission is failing.

Ultimately, we have to fix this by using decode("utf-8")

**Comment by rtgdk** [comment]
@goneall Have opened #272 to track this issue. I opened the PR #273  to fix this issue instance. Again, we have to be sure that decode("utf-8") doesn't fail for any other case - like unicode string, non-ascii characters, etc. Also, we need to make our database consistent in terms of type of data we store. That license was most probably created using python2, that's why it was saved in bytes format. 
Maybe we should revisit the models structure and make sure that only string type is used everywhere for any new data that comes in.
@Ugtan Let us know if you any other ideas/something we might have missed.

File: requirements.txt [hunk 1]
```diff
@@ -1,10 +1,16 @@
-Django==2.1.2
-djangorestframework==3.8.2
-JPype1==0.6.3
-pytz==2018.5
+Django==2.2.14
+jpype1==1.0.1
+numpy==1.19.0
```
Comment (Thrameos) [inline requirements.txt:3]: You should likely upgrade to JPype 1.2 as 1.0 and 1.1 both have a memory leak in strings.

**Comment by rtgdk** [comment]
Thanks for the heads up @Thrameos

**Comment by rtgdk** [comment]
Merging to python3 branch for @Ugtan to work on top of these changes. We will try no more changes go into master. python3 will become master once project completes and master will be changes to python2 to support old version.

Manual testing done.
**Test coverage**
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-b2f8021d-7fff-d89c-d83d-319731181100">&nbsp;
&nbsp; | Total no of tests | Passed
-- | -- | --
app.tests | 86 | 46
api.tests.ValidateFileUploadTests | 2 | 2
api.tests.SubmitLicenseModelsTests | 5 | 1
api.tests.CheckLicenseFileUploadTests | 2 | 1
api.tests.CompareFileUploadTests | 2 | -
api.tests.ConvertFileUploadTests | 8 | -

</b>

@goneall Thanks for the contribution.

**Pull request merged**